File: pyramid_gen/src/pyramid_gen/manifest.py
"""Manifest schema and IO for FITS pyramids.

The manifest is a small JSON sidecar describing every level of a pyramid. It is
a *convenience* index for the browser client: the client must still treat each
.fits.fz file as self-describing and verify the actual compression type from
the file's ZCMPTYPE keyword rather than trusting the manifest's hint.
"""

from __future__ import annotations

import dataclasses
import json
from dataclasses import dataclass, field
from pathlib import Path

MANIFEST_VERSION = 1


@dataclass
class LevelInfo:
    """Metadata for a single pyramid level."""

    z: int
    filename: str
    compression: str  # "GZIP_2" or "RICE_1" -- client hint, verify via ZCMPTYPE
    lossless: bool
    shape: list[int]  # [H, W]
    fpack_tile_count: list[int]  # [n_tiles_y, n_tiles_x]
    pixel_scale_arcsec: float
    wcs: dict  # FITS WCS header as a flat {keyword: value} dict

    def to_dict(self) -> dict:
        return dataclasses.asdict(self)

    @classmethod
    def from_dict(cls, d: dict) -> "LevelInfo":
        return cls(
            z=d["z"],
            filename=d["filename"],
            compression=d["compression"],
            lossless=d["lossless"],
            shape=list(d["shape"]),
            fpack_tile_count=list(d["fpack_tile_count"]),
            pixel_scale_arcsec=d["pixel_scale_arcsec"],
            wcs=dict(d["wcs"]),
        )


@dataclass
class Manifest:
    """Top-level pyramid manifest."""

    source_file: str
    native_shape: list[int]  # [H, W]
    n_levels: int  # N, the deepest z index (so there are N+1 levels)
    levels: list[LevelInfo] = field(default_factory=list)
    fpack_tile_size: int = 256
    version: int = MANIFEST_VERSION

    def to_dict(self) -> dict:
        # Field order chosen to match the documented schema for readability.
        return {
            "version": self.version,
            "source_file": self.source_file,
            "native_shape": list(self.native_shape),
            "fpack_tile_size": self.fpack_tile_size,
            "n_levels": self.n_levels,
            "levels": [lvl.to_dict() for lvl in self.levels],
        }

    @classmethod
    def from_dict(cls, d: dict) -> "Manifest":
        return cls(
            version=d.get("version", MANIFEST_VERSION),
            source_file=d["source_file"],
            native_shape=list(d["native_shape"]),
            fpack_tile_size=d.get("fpack_tile_size", 256),
            n_levels=d["n_levels"],
            levels=[LevelInfo.from_dict(x) for x in d["levels"]],
        )


def write_manifest(path: str | Path, manifest: Manifest) -> None:
    """Serialize a manifest to JSON on disk."""
    path = Path(path)
    with path.open("w") as f:
        json.dump(manifest.to_dict(), f, indent=2, sort_keys=False)
        f.write("\n")


def read_manifest(path: str | Path) -> Manifest:
    """Load a manifest from JSON on disk."""
    path = Path(path)
    with path.open() as f:
        return Manifest.from_dict(json.load(f))

File: fits-pyramid/src/manifest.ts
/**
 * Pyramid manifest schema and loader.
 *
 * The manifest is a small JSON sidecar written by the Phase 1 pipeline. It is a
 * convenience index: the client still treats each `.fits.fz` file as
 * self-describing and verifies the real compression type from its `ZCMPTYPE`
 * keyword. We mirror the Phase 1 schema (`pyramid_gen/manifest.py`).
 */

export interface LevelInfo {
  z: number;
  filename: string;
  /** "GZIP_2" or "RICE_1" — a hint; the file's ZCMPTYPE is authoritative. */
  compression: string;
  lossless: boolean;
  /** [H, W] */
  shape: [number, number];
  /** [n_tiles_y, n_tiles_x] */
  fpack_tile_count: [number, number];
  pixel_scale_arcsec: number;
  /** Flat FITS WCS header as {keyword: value}. */
  wcs: Record<string, unknown>;
}

export interface Manifest {
  version: number;
  source_file: string;
  /** [H, W] */
  native_shape: [number, number];
  fpack_tile_size: number;
  /** N, the deepest z index (the pyramid has N+1 levels, z = 0..N). */
  n_levels: number;
  levels: LevelInfo[];
}

function isObject(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v);
}

function asPair(v: unknown, what: string): [number, number] {
  if (!Array.isArray(v) || v.length !== 2 || typeof v[0] !== 'number' || typeof v[1] !== 'number') {
    throw new Error(`manifest: ${what} must be a [number, number] pair`);
  }
  return [v[0], v[1]];
}

/** Validate and normalize a parsed manifest object. Throws on structural errors. */
export function validateManifest(raw: unknown): Manifest {
  if (!isObject(raw)) throw new Error('manifest: top-level value is not an object');
  if (!Array.isArray(raw.levels)) throw new Error('manifest: "levels" must be an array');

  const levels: LevelInfo[] = raw.levels.map((lvl, idx) => {
    if (!isObject(lvl)) throw new Error(`manifest: level ${idx} is not an object`);
    if (typeof lvl.filename !== 'string') {
      throw new Error(`manifest: level ${idx} is missing a string "filename"`);
    }
    if (typeof lvl.z !== 'number') {
      throw new Error(`manifest: level ${idx} is missing a numeric "z"`);
    }
    return {
      z: lvl.z,
      filename: lvl.filename,
      compression: typeof lvl.compression === 'string' ? lvl.compression : '',
      lossless: lvl.lossless === true,
      shape: asPair(lvl.shape, `level ${idx} shape`),
      fpack_tile_count: asPair(lvl.fpack_tile_count, `level ${idx} fpack_tile_count`),
      pixel_scale_arcsec:
        typeof lvl.pixel_scale_arcsec === 'number' ? lvl.pixel_scale_arcsec : NaN,
      wcs: isObject(lvl.wcs) ? lvl.wcs : {},
    };
  });

  if (levels.length === 0) throw new Error('manifest: "levels" is empty');

  return {
    version: typeof raw.version === 'number' ? raw.version : 1,
    source_file: typeof raw.source_file === 'string' ? raw.source_file : '',
    native_shape: asPair(raw.native_shape, 'native_shape'),
    fpack_tile_size: typeof raw.fpack_tile_size === 'number' ? raw.fpack_tile_size : 256,
    n_levels: typeof raw.n_levels === 'number' ? raw.n_levels : levels.length - 1,
    levels,
  };
}

/** Resolve a level filename to an absolute URL relative to the manifest URL. */
export function resolveLevelUrl(manifestUrl: string, filename: string): string {
  return new URL(filename, manifestUrl).toString();
}

/** Fetch and validate a manifest. */
export async function loadManifest(
  manifestUrl: string,
  fetchImpl: typeof fetch = fetch,
): Promise<Manifest> {
  const resp = await fetchImpl(manifestUrl);
  if (!resp.ok) {
    throw new Error(
      `manifest fetch failed: ${resp.status} ${resp.statusText} for ${manifestUrl}`,
    );
  }
  const json: unknown = await resp.json();
  return validateManifest(json);
}
